### PR TITLE
[refactor] Convert data_files and total_particles into properties

### PR DESCRIPTION
**Pull request by brittonsmith**
## PR Summary

This converts two attributes of `ParticleIndex`, `data_files` and `total_particles`, into properties that call `ParticleIndex._setup_filenames` under the hood. This reduces a little bit of code duplication and should eliminate any instances of `_setup_filenames` being called more than once.

## PR Checklist

- [ ] pass `black --check yt/`
- [ ] pass `isort . --check --diff`
- [ ] pass `flake8 yt/`
- [ ] pass `flynt yt/ --fail-on-change --dry-run -e yt/extern`
- [ ] New features are documented, with docstrings and narrative docs
- [ ] Adds a test for any bugs fixed. Adds tests for new features.

### CODE REVIEW

**Comment by neutrinoceros** [comment]
related to https://github.com/yt-project/yt/issues/2708

**Comment by brittonsmith** [comment]
> I love it ! I'm curious if the property setter pattern actually works as intended with `list.append`, but that's something I trust the test suite to discover :D

Thanks! That's a great question about `append`. I bet it will fail. I'll fix it up if it does.

**Comment by neutrinoceros** [comment]
I _does_ work. Couldn't believe that Travis ran without flinching so I tested this pattern on a dummy class and that's actually valid !
Python is just _that_ magic, seems to me 😅

**Comment by neutrinoceros** [comment]
@brittonsmith you're out of luck, conflicts are already showing up :(

**Comment by brittonsmith** [comment]
> @brittonsmith you're out of luck, conflicts are already showing up :(

This is because PR #2738 was merged. I've never been happier to fix a merge conflict. Just a minute.

File: yt/frontends/halo_catalog/data_structures.py [hunk 1]
```diff
@@ -19,26 +19,10 @@
...
-                    self.io,
-                    self.dataset.parameter_filename,
-                    0,
-                    range=None,
-                )
-            ]
+class HaloCatalogFile(ParticleFile):
+    """
+    Base class for data files of halo catalog datasets.
+>>>>>>> master
```
Comment (neutrinoceros) [inline yt/frontends/halo_catalog/data_structures.py:25]: careful there !

Comment (brittonsmith) [reply yt/frontends/halo_catalog/data_structures.py:25]: oops, didn't commit!

**Comment by neutrinoceros** [comment]
I'll merge when tests are done :)

**Comment by brittonsmith** [comment]
Forgot to remove the `hasattr` line I added in PR #2738. Of course that causes a recursion runaway now. Should be all fixed now.

**Comment by brittonsmith** [comment]
Apparently, I'm getting a flake8 failure for an unrelated file. Does anyone know what's happening there?

**Comment by neutrinoceros** [comment]
I got it too, but I have no idea why we're only seeing it now

**Comment by neutrinoceros** [comment]
Thanks for fixing it here, I wasn't too happy about opening a dedicated PR for this

**Comment by brittonsmith** [comment]
Ok, I think we're good to go here.

**Pull request merged**